### PR TITLE
[Zero-Dim] Scatter gather 0d support

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Others

### PR changes
Others

### Describe
Scatter gather 0d support

#### XPU UT
image

#### NPU/MLU PR
https://github.com/PaddlePaddle/PaddleCustomDevice/pull/243

#### Chinese doc PR
https://github.com/PaddlePaddle/docs/pull/5479

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 1]
```diff
@@ -598,6 +598,26 @@ def test_searchsorted(self):
         self.assertEqual(out.shape, [])
         self.assertEqual(out.numpy(), 0)
 
+    def test_gather(self):
+        x = paddle.to_tensor([1.0, 3.0, 5.0, 7.0, 9.0], stop_gradient=False)
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:602]: 输入x=1D，index=0D的话out会降维为0D，
还应该补个case：输入x>=2D，index=0D，out降维为(x-1) D

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 2]
```diff
@@ -598,6 +598,26 @@ def test_searchsorted(self):
...
+        index = paddle.full([], 2, 'int64')
+        out = paddle.gather(x, index)
+        out.backward()
+
+        self.assertEqual(out.shape, [])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:607]: 和gather一样

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py (hunk 1 above)
Comment (FeixLiu) [reply python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:602]: done

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 3]
```diff
@@ -598,6 +598,61 @@ def test_searchsorted(self):
...
+    def test_gather_xD_axis_1(self):
+        x = paddle.to_tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+        index = paddle.full([], 1, 'int64')
+        out = paddle.gather(x, index, axis=1)
+
+        self.assertEqual(out.shape, [2])
+        for i in range(2):
+            self.assertEqual(out.numpy()[i], x.numpy()[i][1])
+
+    def test_scatter_0D(self):
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py]: test_scatter_1D

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 4]
```diff
@@ -598,6 +598,61 @@ def test_searchsorted(self):
...
+        x = paddle.to_tensor(
+            [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], stop_gradient=False
+        )
+        index = paddle.full([], 1, 'int64')
+        out = paddle.gather(x, index)
+        out.backward()
+
+        self.assertEqual(out.shape, [3])
+        for i in range(3):
+            self.assertEqual(out.numpy()[i], x.numpy()[1][i])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:621]: 建议写成 x.numpy()[1, i]

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 5]
```diff
@@ -598,6 +598,61 @@ def test_searchsorted(self):
...
+        self.assertEqual(out.grad.shape, [3])
+
+    def test_gather_xD_axis_1(self):
+        x = paddle.to_tensor([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
+        index = paddle.full([], 1, 'int64')
+        out = paddle.gather(x, index, axis=1)
+
+        self.assertEqual(out.shape, [2])
+        for i in range(2):
+            self.assertEqual(out.numpy()[i], x.numpy()[i][1])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:631]: 建议写成 x.numpy()[i, 1]

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 6]
```diff
@@ -598,6 +598,61 @@ def test_searchsorted(self):
...
+        x = paddle.to_tensor(
+            [[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]], stop_gradient=False
+        )
+        index = paddle.full([], 1, 'int64')
+        updates = paddle.to_tensor([1.0, 2.0, 3.0])
+        out = paddle.scatter(x, index, updates)
+        out.backward()
+
+        for i in range(3):
+            self.assertEqual(out.numpy()[1][i], updates.numpy()[i])
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:653]: 建议写成 out.numpy()[1, i]

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py [hunk 7]
```diff
@@ -679,6 +734,68 @@ def test_searchsorted(self):
...
+        x = paddle.full([2, 3], 1.0, 'float32')
+        index = paddle.full([], 1, 'int64')
+        out = paddle.gather(x, index)
+        paddle.static.append_backward(out)
+
+        prog = paddle.static.default_main_program()
+        res = self.exe.run(prog, fetch_list=[out])
+        self.assertEqual(res[0].shape, (3,))
+        for i in range(3):
+            self.assertEqual(res[0][i], 1)
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py:760]: 同上，建议把切片形式写在一个[]里，调一次numpy切片API

**Review by zhwesky2010 — COMMENTED** [review]
API文档也改下吧
infoflow 2022-12-02 17-09-16
infoflow 2022-12-02 17-12-40

File: python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py (hunk 3 above)
Comment (FeixLiu) [reply python/paddle/fluid/tests/unittests/test_zero_dim_tensor.py]: done

**Review by zhwesky2010 — APPROVED** [review]
LGTM

**Review by Ligoml — APPROVED** [review]
LGTM for docs，中文文档也同步更新一下吧

**Pull request merged**